cluster/gce/windows/k8s-node-setup.psm1

```diff
@@ -1668,6 +1668,11 @@ function Configure-LoggingAgent {
   $config = $FLUENTD_CONFIG.replace('NODE_NAME', (hostname))
   $config | Out-File -FilePath $fluentd_config_file -Encoding ASCII
   Log-Output "Wrote fluentd logging config to $fluentd_config_file"
+
+  # Configure StackdriverLogging to automatically restart on failure after 10
+  # seconds. The logging agent may die die to various disruptions but can be
+  # resumed.
```

Comment:
If a failure occurs, do you know what happens to the log file that the agent writes to? When the agent restarts will the log messages that would contain information about the failure be preserved, or will the agent clear the old logs?

If the latter, it would be helpful to see if we can figure out a way to preserve the failure logs.

Reply:
I booted up a Windows and ran `taskkill /F /IM ruby.exe`.

You can see this in the logs StackDriver logs:
```
The Stackdriver Logging service terminated unexpectedly.  It has done this 1 time(s).  The following corrective action will be taken in 1000 milliseconds: Restart the service.

The Stackdriver Logging service entered the running state.
```

Reply:
Thanks for checking that. There's also a log file at `C:\programfilesx86\stackdriver\fluentd.log`, or something like that (pulled the path from an old chat). Can you check that file too? 

I'm not sure if fluentd is configured to send its own log file to Stackdriver logging - if it is, cool. If not, it would be nice if the file continues to get appended to after the restart, rather than overwritten.

Reply:
The logs are preserved.
`type 'C:\Program Files (x86)\Stackdriver\LoggingAgent\fluentd.log'`


```
2020-08-11 21:32:19 +0000 [info]: starting fluentd-1.7.4 pid=1300 ruby="2.5.5"
2020-08-11 21:32:19 +0000 [info]: spawn command to main:  cmdline=["C:/Program Files (x86)/Stackdriver/LoggingAgent/Main/bin/ruby.exe", "-Eascii-8bit:ascii-8bit", "C:/Program Files (x86)/Stackdriver/LoggingAgent/Main/bin/fluentd", "-c", "C:\\Program Files (x86)\\Stackdriver\\LoggingAgent\\fluent.conf", "-o", "C:\\Program Files (x86)\\Stackdriver\\LoggingAgent\\fluentd.log", "-x", "StackdriverLogging", "--under-supervisor"]
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-detect-exceptions' version '0.0.12'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-google-cloud' version '0.8.1'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-kubernetes_metadata_filter' version '2.4.1'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-multi-format-parser' version '1.0.0'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-parser-winevt_xml' version '0.2.0'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-prometheus' version '1.4.0'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-record-modifier' version '2.0.1'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-record-reformer' version '0.9.1'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '2.2.0'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-s3' version '1.1.10'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-systemd' version '1.0.2'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-webhdfs' version '1.2.3'
2020-08-11 21:32:22 +0000 [info]: gem 'fluent-plugin-windows-eventlog' version '0.4.4'
2020-08-11 21:32:22 +0000 [info]: gem 'fluentd' version '1.7.4'
2020-08-11 21:32:22 +0000 [info]: gem 'fluentd' version '1.6.3'
2020-08-11 21:32:22 +0000 [info]: adding match pattern="reform.**" type="record_reformer"
2020-08-11 21:32:22 +0000 [info]: adding match pattern="{raw.stderr,raw.stdout}" type="google_cloud"
2020-08-11 21:32:23 +0000 [info]: #0 Detected GCE platform
2020-08-11 21:32:23 +0000 [info]: #0 Logs viewer address: https://console.cloud.google.com/logs/viewer?project=REDACTED
2020-08-11 21:32:23 +0000 [info]: adding filter pattern="**" type="record_transformer"
2020-08-11 21:32:24 +0000 [info]: adding filter pattern="winevt.raw" type="record_transformer"
2020-08-11 21:32:24 +0000 [info]: adding filter pattern="**" type="add_insert_ids"
2020-08-11 21:32:24 +0000 [info]: adding match pattern="**" type="google_cloud"
2020-08-11 21:32:24 +0000 [warn]: #0 'partial_success' parameter is deprecated: This feature is permanently enabled
2020-08-11 21:32:24 +0000 [info]: #0 Detected GCE platform
2020-08-11 21:32:24 +0000 [info]: #0 Logs viewer address: https://console.cloud.google.com/logs/viewer?project=REDACTED
2020-08-11 21:32:24 +0000 [info]: adding source type="tail"
2020-08-11 21:32:24 +0000 [info]: adding source type="tail"
2020-08-11 21:32:24 +0000 [info]: adding source type="tail"
2020-08-11 21:32:24 +0000 [info]: adding source type="tail"
2020-08-11 21:32:24 +0000 [info]: adding source type="windows_eventlog"
2020-08-11 21:32:24 +0000 [info]: #0 starting fluentd worker pid=4272 ppid=1300 worker=0
2020-08-11 21:32:24 +0000 [info]: #0 Started the add_insert_ids plugin with logging.googleapis.com/insertId as the insert ID key.
2020-08-11 21:32:24 +0000 [info]: #0 Initialized the insert ID key to REDACTED.
2020-08-11 21:32:24 +0000 [info]: #0 following tail of /etc/kubernetes/logs/kube-proxy.log
2020-08-11 21:32:24 +0000 [info]: #0 following tail of /etc/kubernetes/logs/kubelet.log
2020-08-11 21:32:24 +0000 [info]: #0 fluentd worker is now running worker=0
2020-08-11 21:32:24 +0000 [info]: #0 disable filter chain optimization because [Fluent::Plugin::RecordTransformerFilter] uses `#filter_stream` method.
2020-08-11 21:32:26 +0000 [info]: #0 disable filter chain optimization because [Fluent::Plugin::RecordTransformerFilter, Fluent::Plugin::RecordTransformerFilter] uses `#filter_stream` method.
2020-08-11 21:32:28 +0000 [info]: #0 disable filter chain optimization because [Fluent::Plugin::RecordTransformerFilter] uses `#filter_stream` method.
2020-08-11 21:32:29 +0000 [info]: #0 disable filter chain optimization because [Fluent::Plugin::RecordTransformerFilter] uses `#filter_stream` method.
2020-08-11 21:32:29 +0000 [info]: #0 Successfully sent gRPC to Stackdriver Logging API.
2020-08-11 23:50:29 +0000 [info]: parsing config file is succeeded path="C:\\Program Files (x86)\\Stackdriver\\LoggingAgent\\fluent.conf"
2020-08-11 23:50:33 +0000 [info]: Detected GCE platform
2020-08-11 23:50:33 +0000 [info]: Logs viewer address: https://console.cloud.google.com/logs/viewer?project=REDACTED
2020-08-11 23:50:33 +0000 [warn]: 'partial_success' parameter is deprecated: This feature is permanently enabled
2020-08-11 23:50:33 +0000 [info]: Detected GCE platform
2020-08-11 23:50:33 +0000 [info]: Logs viewer address: https://console.cloud.google.com/logs/viewer?project=REDACTED
2020-08-11 23:50:34 +0000 [info]: using configuration file: <ROOT>
  <source>
    @type tail
    path "/var/log/containers/*.log"
    pos_file "/var/log/gcp-containers.log.pos"
    tag "reform.*"
    read_from_head true
    <parse>
      @type "multi_format"
      <pattern>
        format json
        time_key "time"
        time_format "%Y-%m-%dT%H:%M:%S.%NZ"
        time_type string
      </pattern>
      <pattern>
        format /^(?<time>.+) (?<stream>stdout|stderr) [^ ]* (?<log>.*)$/
        time_format "%Y-%m-%dT%H:%M:%S.%N%:z"
        expression ^(?<time>.+) (?<stream>stdout|stderr) [^ ]* (?<log>.*)$
        ignorecase false
        multiline false
      </pattern>
    </parse>
  </source>
  <source>
    @type tail
    format multiline
    multiline_flush_interval 5s
    format_firstline /^\w\d{4}/
    format1 /^(?<severity>\w)(?<time>\d{4} [^\s]*)\s+(?<pid>\d+)\s+(?<source>[^ \]]+)\] (?<message>.*)/
    time_format %m%d %H:%M:%S.%N
    path "/etc/kubernetes/logs/kubelet.log"
    pos_file "/etc/kubernetes/logs/gcp-kubelet.log.pos"
    tag "kubelet"
    <parse>
      time_format %m%d %H:%M:%S.%N
      format_firstline /^\w\d{4}/
      @type multiline
      format1 /^(?<severity>\w)(?<time>\d{4} [^\s]*)\s+(?<pid>\d+)\s+(?<source>[^ \]]+)\] (?<message>.*)/
    </parse>
  </source>
  <source>
    @type tail
    format multiline
    multiline_flush_interval 5s
    format_firstline /^\w\d{4}/
    format1 /^(?<severity>\w)(?<time>\d{4} [^\s]*)\s+(?<pid>\d+)\s+(?<source>[^ \]]+)\] (?<message>.*)/
    time_format %m%d %H:%M:%S.%N
    path "/etc/kubernetes/logs/kube-proxy.log"
    pos_file "/etc/kubernetes/logs/gcp-kube-proxy.log.pos"
    tag "kube-proxy"
    <parse>
      time_format %m%d %H:%M:%S.%N
      format_firstline /^\w\d{4}/
      @type multiline
      format1 /^(?<severity>\w)(?<time>\d{4} [^\s]*)\s+(?<pid>\d+)\s+(?<source>[^ \]]+)\] (?<message>.*)/
    </parse>
  </source>
  <source>
    @type tail
    format multiline
    multiline_flush_interval 5s
    format_firstline /^time=/
    format1 /^time="(?<time>[^ ]*)" level=(?<severity>\w*) (?<message>.*)/
    time_format %Y-%m-%dT%H:%M:%S.%N%z
    path "/etc/kubernetes/logs/containerd.log"
    pos_file "/etc/kubernetes/logs/gcp-containerd.log.pos"
    tag "container-runtime"
    <parse>
      time_format %Y-%m-%dT%H:%M:%S.%N%z
      format_firstline /^time=/
      @type multiline
      format1 /^time="(?<time>[^ ]*)" level=(?<severity>\w*) (?<message>.*)/
    </parse>
  </source>
  <match reform.**>
    @type record_reformer
    enable_ruby true
    tag "${if record[\'stream\'] == \'stderr\' then \'raw.stderr\' else \'raw.stdout\' end}"
    remove_keys "stream,log"
    <record>
      logging.googleapis.com/local_resource_id ${"k8s_container.#{tag_suffix[4].rpartition('.')[0].split('_')[1]}.#{tag_suffix[4].rpartition('.')[0].split('_')[0]}.#{tag_suffix[4].rpartition('.')[0].split('_')[2].rpartition('-')[0]}"}
      message ${record['log']}
      severity ${record['severity'] || if record['stream'] == 'stderr' then 'ERROR' else 'INFO' end}
    </record>
  </match>
  <match {raw.stderr,raw.stdout}>
    @type google_cloud
    detect_json true
    split_logs_by_tag false
    buffer_type "file"
    buffer_path /var/log/fluentd-buffers/kubernetes.containers.buffer
    buffer_queue_full_action block
    buffer_chunk_limit 512k
    buffer_queue_limit 6
    flush_interval 5s
    max_retry_wait 30
    disable_retry_limit
    num_threads 2
    use_grpc true
    adjust_invalid_timestamps false
    <buffer>
      flush_mode interval
      retry_type exponential_backoff
      @type file
      path /var/log/fluentd-buffers/kubernetes.containers.buffer
      flush_thread_count 2
      flush_interval 5s
      retry_forever
      retry_max_interval 30
      chunk_limit_size 512k
      queue_limit_length 6
      overflow_action block
    </buffer>
  </match>
  <filter **>
    @type record_transformer
    enable_ruby true
    <record>
      logging.googleapis.com/local_resource_id ${"k8s_node.kubernetes-windows-node-group-m5zm"}
    </record>
  </filter>
  <filter winevt.raw>
    @type record_transformer
    enable_ruby
    <record>
      message ${record['description']}
      severity ${record['event_type'] && record['event_type'].length > 0 && !record['event_type']['audit'] ? record['event_type'][0].upcase : ''}
      timestamp ${t = Time.parse(record['time_generated']); {'seconds' => t.tv_sec, 'nanos' => t.tv_nsec}}
    </record>
  </filter>
  <filter **>
    @type add_insert_ids
  </filter>
  <match **>
    @type google_cloud
    buffer_type "file"
    buffer_path /var/log/google-fluentd/buffers
    buffer_chunk_limit 512KB
    flush_interval 5s
    disable_retry_limit false
    retry_limit 3
    retry_wait 10
    max_retry_wait 300
    num_threads 8
    use_grpc true
    partial_success true
    <buffer>
      flush_mode interval
      retry_type exponential_backoff
      @type file
      path /var/log/google-fluentd/buffers
      flush_thread_count 8
      flush_interval 5s
      retry_forever false
      retry_max_times 3
      retry_max_interval 300
      chunk_limit_size 512KB
    </buffer>
  </match>
  <source>
    @type windows_eventlog
    channels application,system,security
    read_interval 2
    tag "winevt.raw"
    <storage>
      @type "local"
      persistent true
      path "C:\\Program Files (x86)\\Stackdriver\\LoggingAgent\\Main\\pos\\winevtlog.pos"
    </storage>
  </source>
</ROOT>
2020-08-11 23:50:34 +0000 [info]: starting fluentd-1.7.4 pid=3120 ruby="2.5.5"
2020-08-11 23:50:34 +0000 [info]: spawn command to main:  cmdline=["C:/Program Files (x86)/Stackdriver/LoggingAgent/Main/bin/ruby.exe", "-Eascii-8bit:ascii-8bit", "C:/Program Files (x86)/Stackdriver/LoggingAgent/Main/bin/fluentd", "-c", "C:\\Program Files (x86)\\Stackdriver\\LoggingAgent\\fluent.conf", "-o", "C:\\Program Files (x86)\\Stackdriver\\LoggingAgent\\fluentd.log", "-x", "StackdriverLogging", "--under-supervisor"]
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-detect-exceptions' version '0.0.12'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-google-cloud' version '0.8.1'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-kubernetes_metadata_filter' version '2.4.1'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-multi-format-parser' version '1.0.0'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-parser-winevt_xml' version '0.2.0'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-prometheus' version '1.4.0'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-record-modifier' version '2.0.1'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-record-reformer' version '0.9.1'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '2.2.0'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-s3' version '1.1.10'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-systemd' version '1.0.2'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-webhdfs' version '1.2.3'
2020-08-11 23:50:37 +0000 [info]: gem 'fluent-plugin-windows-eventlog' version '0.4.4'
2020-08-11 23:50:37 +0000 [info]: gem 'fluentd' version '1.7.4'
2020-08-11 23:50:37 +0000 [info]: gem 'fluentd' version '1.6.3'
2020-08-11 23:50:37 +0000 [info]: adding match pattern="reform.**" type="record_reformer"
2020-08-11 23:50:37 +0000 [info]: adding match pattern="{raw.stderr,raw.stdout}" type="google_cloud"
2020-08-11 23:50:38 +0000 [info]: #0 Detected GCE platform
2020-08-11 23:50:39 +0000 [info]: #0 Logs viewer address: https://console.cloud.google.com/logs/viewer?project=REDACTED
2020-08-11 23:50:39 +0000 [info]: adding filter pattern="**" type="record_transformer"
2020-08-11 23:50:39 +0000 [info]: adding filter pattern="winevt.raw" type="record_transformer"
2020-08-11 23:50:39 +0000 [info]: adding filter pattern="**" type="add_insert_ids"
2020-08-11 23:50:39 +0000 [info]: adding match pattern="**" type="google_cloud"
2020-08-11 23:50:39 +0000 [warn]: #0 'partial_success' parameter is deprecated: This feature is permanently enabled
2020-08-11 23:50:39 +0000 [info]: #0 Detected GCE platform
2020-08-11 23:50:39 +0000 [info]: #0 Logs viewer address: https://console.cloud.google.com/logs/viewer?project=REDACTED
2020-08-11 23:50:39 +0000 [info]: adding source type="tail"
2020-08-11 23:50:39 +0000 [info]: adding source type="tail"
2020-08-11 23:50:39 +0000 [info]: adding source type="tail"
2020-08-11 23:50:39 +0000 [info]: adding source type="tail"
2020-08-11 23:50:39 +0000 [info]: adding source type="windows_eventlog"
2020-08-11 23:50:39 +0000 [info]: #0 starting fluentd worker pid=4276 ppid=3120 worker=0
2020-08-11 23:50:39 +0000 [info]: #0 Started the add_insert_ids plugin with logging.googleapis.com/insertId as the insert ID key.
2020-08-11 23:50:39 +0000 [info]: #0 Initialized the insert ID key to REDACTED.
2020-08-11 23:50:39 +0000 [info]: #0 following tail of /etc/kubernetes/logs/kube-proxy.log
2020-08-11 23:50:39 +0000 [info]: #0 following tail of /etc/kubernetes/logs/kubelet.log
2020-08-11 23:50:39 +0000 [info]: #0 disable filter chain optimization because [Fluent::Plugin::RecordTransformerFilter] uses `#filter_stream` method.
2020-08-11 23:50:39 +0000 [info]: #0 fluentd worker is now running worker=0
2020-08-11 23:50:39 +0000 [info]: #0 disable filter chain optimization because [Fluent::Plugin::RecordTransformerFilter] uses `#filter_stream` method.
2020-08-11 23:50:41 +0000 [info]: #0 disable filter chain optimization because [Fluent::Plugin::RecordTransformerFilter, Fluent::Plugin::RecordTransformerFilter] uses `#filter_stream` method.
2020-08-11 23:50:45 +0000 [info]: #0 Successfully sent gRPC to Stackdriver Logging API.
2020-08-11 23:50:47 +0000 [info]: #0 disable filter chain optimization because [Fluent::Plugin::RecordTransformerFilter] uses `#filter_stream` method.
```